components/num-util/src/fixed_decimal.rs

```diff
@@ -373,6 +377,135 @@ impl fmt::Display for FixedDecimal {
+        for (i, c) in no_sign_str.chars().enumerate() {
+            if c == '.' {
+                match has_dot {
+                    false => {
+                        dot_index = i;
+                        has_dot = true;
+                        if i == 0 || i == no_sign_str.len() - 1 {
+                            return Err(Error::Syntax);
+                        }
+                    }
+                    true => {
+                        return Err(Error::Syntax);
+                    }
+                }
+            } else if !c.is_digit(10) {
+                return Err(Error::Syntax);
+            }
+        }
+        // no_dot_str: shows the string when the dot is removed from it
+        let no_dot_str = no_sign_str.replace(".", "");
+        let no_dot_str_len = no_dot_str.chars().count();
+
+        // left_zeros: number of zeros at the left of no_dot_str
+        // right_zeros: number of zeros at the right of no_dot_str
+        // These are not necessarily leading or trailing zeros
+        // Example:
+        //     input string     left_zeros      right_zeros
+        //     00123000             2                3
+        //     0.0123000            2               3
+        //     001.23000            2               3
+        //     001230.00            2               3
+        let mut left_zeros = 0;
+        for (_i, c) in no_dot_str.chars().enumerate() {
+            if c == '0' {
+                left_zeros += 1;
+            } else {
+                break;
+            }
+        }
+        // If the input only has 0 (like 000, 00.0, -00.000) we handle the situation here
+        // and return the correct FixedDecimal now and don't run the rest of the code
+        if left_zeros == no_dot_str_len {
+            let mut dec: FixedDecimal = 0.into();
+            dec.is_negative = is_negative;
+            dec.upper_magnitude = left_zeros as i16 - 1i16;
+            if has_dot {
+                dec.upper_magnitude -= no_dot_str_len as i16 - dot_index as i16;
+                dec.lower_magnitude -= no_dot_str_len as i16 - dot_index as i16;
+            }
+            return Ok(dec);
+        }
+        let mut right_zeros = 0;
+        for (_i, c) in no_dot_str.chars().rev().enumerate() {
+            if c == '0' {
+                right_zeros += 1;
+            } else {
+                break;
+            }
+        }
+        // digits_str: the string after removing sign, dot, left zeros, and right zeros
+        let digits_str = &no_dot_str[left_zeros..no_dot_str_len - right_zeros];
+
+        // shift: number of digits after dot in the input_str, 0 if there is no dot
+        let mut shift: i16 = 0;
+        if has_dot {
+            shift = (no_sign_str_len - dot_index - 1) as i16;
+        }
+
+        // Constructing DecimalFixed only using digits_str
+        let mut dec: FixedDecimal = 0.into();
+        let mut v = SmallVec::<[u8; 8]>::new();
```

Comment:
You can pre-allocate the capacity:

https://docs.rs/smallvec/1.4.2/smallvec/struct.SmallVec.html#method.with_capacity

The length of the smallvec should end up being `digits_str.len()`.  When you finish appending all the digits, you can `debug_assert!` that this is indeed the case.